### DISCUSSION TITLE
Architecture: split backend and frontend - 2 dockers

### DISCUSSION THREAD

**Discussion by shuther** [danny-avila/LibreChat#1299] [Help Wanted]
I would like to split the API and the front end, and get them running under 2 different dockers because I need to add a middleware.
I found an example for [Heroku](https://docs.librechat.ai/deployment/heroku.html) as well as the one in the [dev folder](https://github.com/danny-avila/LibreChat/blob/0bae503a0ab99ee71b367a11f2a79e5663262420/docs/dev/deploy-compose.yml) but I had no luck:
1. Where in the frontend could I specify the API endpoint (URL/port)?
2. I am expecting to run the API and the frontend on 2 different ports. 
For the API, I created this Dockerfile:
```
FROM node:19-alpine AS base

ENV PORT=9000

WORKDIR /api
COPY /api/package*.json /api/
WORKDIR /
COPY /config/ /config/
COPY /package*.json /
RUN npm ci

# Node API setup

FROM base AS node-api
WORKDIR /api
COPY /api/ /api/
EXPOSE $PORT
ENV HOST=0.0.0.0
CMD ["npm", "start"]
```
Is it the correct cmd ?
I removed the PORT variable in .env

For the client I have this one:
```
# Base node image
FROM node:19-alpine AS base
WORKDIR /client
COPY /client/package*.json /client/
WORKDIR /
COPY /config/ /config/
COPY /package*.json /

WORKDIR /packages/data-provider
COPY /packages/data-provider ./
RUN npm install && npm run build

WORKDIR /
RUN npm ci

# React client build
FROM base AS react-client
WORKDIR /client
COPY /client/ /client/
ENV NODE_OPTIONS="--max-old-space-size=2048"
RUN npm run build

# Nginx setup
FROM nginx:stable-alpine AS nginx-client
WORKDIR /usr/share/nginx/html
COPY --from=react-client /client/dist /usr/share/nginx/html
COPY client/nginx.conf /etc/nginx/conf.d/default.conf
ENTRYPOINT ["nginx", "-g", "daemon off;"]
```

In the log I have this error:

> LibreChat-frontend  | 2023/12/06 18:04:53 [warn] 1#1: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:2
> LibreChat-frontend  | nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:2
> LibreChat-frontend  | 2023/12/06 18:04:53 [emerg] 1#1: mkdir() "/var/cache/nginx/client_temp" failed (13: Permission denied)
> LibreChat-frontend  | nginx: [emerg] mkdir() "/var/cache/nginx/client_temp" failed (13: Permission denied)
> 

3. In the log I have this:

> LibreChat-backend   | 
> LibreChat-backend   | > @librechat/backend@0.6.1 start
> LibreChat-backend   | > echo 'please run this from the root directory'
> LibreChat-backend   | 
> LibreChat-backend   | 
> LibreChat-backend   | please run this from the root directory
> LibreChat-frontend exited with code 1

Any help would be appreciated.
Thanks

**Comment by danny-avila** [comment]
Have you seen the second Dockerfile at the root of the project?

Dockerfile.multi

note that the “client” is just static code. It runs nothing and the server needs access to it to serve the static files. This is why you would need something like NGINX to configure it separately, which the dockerfile I mentioned does

**Reply by danny-avila** [reply]
This is a guide that uses that dockerfile: https://docs.librechat.ai/deployment/digitalocean.html

**Reply by shuther** [reply]
I think I can leverage Dockerfile.multi; just wondering, is threre a Dockerfile to generate the image: ghcr.io/danny-avila/librechat-dev-api:latest ? Also, maybe you want to mention librechat-api in this docker compose file?

**Reply by demon200slayer** [reply]
@shuther were you able to run backend separately? Guide me.

**Comment by gabrielnocce** [comment]
@shuther  did you solve that?
